src/cargo/ops/vendor.rs

```diff
@@ -387,3 +442,53 @@ fn copy_and_checksum(src_path: &Path, dst_path: &Path, buf: &mut [u8]) -> CargoR
             .with_context(|| format!("failed to write to {:?}", dst_path))?;
     }
 }
+
+/// Maps source id to separate vendor directory name.
+/// E.g. registry+deadbeeffeedcafe, git+0123456789abcdef
+fn source_id_to_dir_name(src_id: SourceId) -> String {
+    let src_type = if src_id.is_registry() {
+        "registry"
+    } else if src_id.is_git() {
+        "git"
+    } else {
+        panic!("Invalid source ID: {src_id}")
+    };
+    struct SourceIdShortHash(SourceId);
+    impl Hash for SourceIdShortHash {
+        fn hash<H: Hasher>(&self, state: &mut H) {
+            self.0.stable_hash(Path::new(""), state);
+        }
+    }
+    let hash = util::short_hash(&SourceIdShortHash(src_id));
+    format!("{src_type}-{hash}")
+}
+
+/// Removes all contents under the given path, except hidden or listed in `keep`.
+fn remove_dir_contents(path: &Path, keep: &HashSet<PathBuf>) -> CargoResult<()> {
```

Comment:
How come this is necessary? Is this solving a new issue? (sorry I just didn't see this before and otherwise didn't expect this PR to be modifying the removal pieces for this feature)

Reply:
This merges several removal logic from these places.

- https://github.com/alexcrichton/cargo-vendor/blob/07570e23d0841936cea8af2a125484f4d10ccdfa/src/main.rs#L402-L408
- https://github.com/rust-lang/cargo/blob/25fcb135d02ea897ce894b67ae021f48107d522b/src/cargo/ops/vendor.rs#L90-L102
- https://github.com/rust-lang/cargo/blob/25fcb135d02ea897ce894b67ae021f48107d522b/src/cargo/ops/vendor.rs#L240-L246